### PR TITLE
Proper float64 support for unary elemwise operators (mshadow_op.h)

### PR DESCRIPTION
**Pull request by mseeger**
Modified mshadow_op.h s.t. elemwise unary operators support dtype float64 properly.
Previous code does everything in float32
- Had to switch from math.h to cmath, and use std:: prefix, in order to avoid naming clashes

### CODE REVIEW

**Comment by mseeger** [comment]
Something is broken, but not in my code. I'd love to get to the unit tests on GPUs, to see whether my float64 code breaks things there ...

File: src/operator/mshadow_op.h [hunk 1]
```diff
@@ -99,22 +99,30 @@ struct reciprocal_grad {
...
   template<typename DType>
   MSHADOW_XINLINE static DType Map(DType a) {
-    return DType(DType(1.0f) / (DType(1.0f) + expf(-a)));
+    return DType(DType(1.0f) / (DType(1.0f) + std::expf(-a)));
   }
 };
+
+template<>
+MSHADOW_XINLINE double sigmoid::Map<double>(double a) {
+  return 1.0 / (1.0 + std::exp(-a));
```
Comment (piiswrong) [inline src/operator/mshadow_op.h]: instead of specializing for each operator, I think it's better to make a templated function

template<typename DType>
mxnet_math::exp(DType a);

and specialize this.

Comment (piiswrong) [reply src/operator/mshadow_op.h]: mxnet_math may not be a good namespace. Maybe mxnet_exp?
I'm open to suggestions

Comment (mseeger) [reply src/operator/mshadow_op.h]: I followed the example of lgamma (due to Valentin Flunkert, I think).

Would it work just to do template versions of the math.h functions?

I am thinking about these DType(...) expressions everywhere.

Comment (piiswrong) [reply src/operator/mshadow_op.h]: `Would it work just to do template versions of the math.h functions?`

What does this mean? If you want to overload std::exp that's not a good idea.

Comment (mseeger) [reply src/operator/mshadow_op.h]: OK, I will be OOTO for a while, so I think this change has to wait (I may delete the PR).

Still, could you be more explicit how you would do it? Here is what I tried:
- Just added the template specializations for double for each XXX:Map
- Problem: Did not work for exp, expm1, log, ..., because now, you have a clash with
   the global names included via math.h
- At this point, I switched to including cmath, so it is std::exp, std::expm1, std::log.
   I agree this is not elegant, because lots of code has to change.

What would you suggest?

I am not even 100% why these name clashes really appear. For example, if I call
exp(a), where a is double, how can it confuse this with

struct exp { ... }

The latter does not have a constructor with double arg.

Sorry I am not so greatly familiar with C++. Maybe there is a much more elegant solution.
If you can suggest one, please tell me.

Comment (piiswrong) [reply src/operator/mshadow_op.h]: We can implement our own set of math functions under the namespace mxnet::math. For example
```
namespace mxnet {
namespace math {
template<typename DType>
DType exp(DType value);

template<>
double exp(double value);

}
}

**Comment by mseeger** [comment]
OK, did as you suggested. Except in two cases:
- Valentin's code: Restricted to float and double
- power_grad: Oddly enough, this is instantiated for DType unsigned char, signed char.
   Sounds like a bit dangerous ...

**Comment by mseeger** [comment]
Darn, this exp(.) and expf(.) business is really broken, compiles on Mac, but not on Ubuntu.

I'll go for the C++11 definition: Always std::exp(.), if cmath is included.

**Comment by mseeger** [comment]
OK, I rewrote this under the premise that computations are done by casting to float and back for all DType's except double. I hope this is what is intended here.

**Comment by mseeger** [comment]
@asmushetzel 
Hi Asmus, could you have a look at this PR? I rewrote src/operator/mshadow_op.h in order to do the right thing for double as well. It does not build on GPUs.

Could this be because I am using cmath instead of math.h, and the C++11 versions?

Note that we need to get this done for the Gaussian process use case.

**Comment by mseeger** [comment]
src/operator/contrib/./../mshadow_op.h(139): error: calling a __host__ function from a __host__ __device__ function is not allowed

?? No idea what this means

**Comment by szha** [comment]
It's failing on cuda compilation, because `math::exp` isn't available on device.

File: src/operator/math_functions-inl.h [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+#define MXNET_OPERATOR_MATH_FUNCTIONS_INL_H_
+
+#include <cmath>
+
+namespace mxnet {
+namespace op {
+
+namespace math {
+
+#define MATH_FUNC1(name) \
```
Comment (piiswrong) [inline src/operator/math_functions-inl.h]: MXNET_UNARY_MATH_FUNC

File: src/operator/math_functions-inl.h [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+template<typename DType> inline \
+DType name(DType a) { \
```
Comment (piiswrong) [inline src/operator/math_functions-inl.h]: Add MSHADOW_XINLINE to make it callable on GPU

File: src/operator/math_functions-inl.h [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  return DType(std::name(static_cast<float>(a))); \
```
Comment (piiswrong) [inline src/operator/math_functions-inl.h]: You can't use std:: function on gpu. Please see cuda manual for correct usage

File: src/operator/math_functions-inl.h [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+template<> inline \
+float name(float a) { \
+  return std::name(a); \
+} \
+template<> inline \
+double name(double a) { \
+  return std::name(a); \
+} \
+template<typename DType> inline \
+float name##f(DType a) { \
```
Comment (piiswrong) [inline src/operator/math_functions-inl.h]: Why do we want this? Looks weird

File: src/operator/mshadow_op.h [hunk 6]
```diff
@@ -141,213 +186,223 @@ struct xelu_grad {
...
   template<typename DType>
   MSHADOW_XINLINE static DType Map(DType a) {
-    return DType(DType(1.0f) - a * a);
+    float af(static_cast<float>(a));
+    return DType(1.0f - af * af);
   }
 };
 
+template<>
+MSHADOW_XINLINE double tanh_grad::Map<double>(double a) {
```
Comment (piiswrong) [inline src/operator/mshadow_op.h]: why is this needed?

File: src/operator/mshadow_op.h [hunk 7]
```diff
@@ -84,41 +102,68 @@ struct negation {
 struct reciprocal {
   template<typename DType>
   MSHADOW_XINLINE static DType Map(DType a) {
-    return DType(1.0f/a);
+    float af(static_cast<float>(a));
```
Comment (piiswrong) [inline src/operator/mshadow_op.h]: why?

File: src/operator/math_functions-inl.h (hunk 4 above)
Comment (mseeger) [reply src/operator/math_functions-inl.h]: OK, I switch back to math.h

File: src/operator/math_functions-inl.h (hunk 5 above)
Comment (mseeger) [reply src/operator/math_functions-inl.h]: The idea is that whenever any math.h function is used in an expression with DType, I compute it internally in float, and cast back to DType in the end. This is why the ***f variants are needed here: inside an expression, you do not want to cast back to DType, only at the end.

File: src/operator/mshadow_op.h (hunk 7 above)
Comment (mseeger) [reply src/operator/mshadow_op.h]: My idea was to do computations in float and cast back to DType in the end.

I see now this is not really the best. Here is what I will do:
- Whenever any math.h function is involved in an expression, I will do computations in
   float and cast to DType in the end
- I'll leave all other expressions (purely arithmetic, no math.h) as they are. This will of
   course result in garbage if DType is not floating point, but maybe this is intended

**Comment by mseeger** [comment]
OK, all tests go through now

**Comment by asmushetzel** [comment]
A lot of tedious work.
We have to agree on what we do with cases like:
int x = 4; float y = foo(x);
This code  will always cast back to an int at the end, so y may be pretty much garbage. Which is in contrast to the usual expectations when you do something  like 
float y = exp(x) 
where you will actually get the same result as if you would do 
float y = float(exp(double(x)));

**Comment by mseeger** [comment]
Well, it is always DType in = DType out, right?

I am happy with the current solution. It reflects the state we had before, but is now doing the right thing for DType = double.

File: src/operator/math_functions-inl.h [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+namespace mxnet {
+namespace op {
+
+namespace math {
+
+#define MXNET_UNARY_MATH_FUNC(name) \
+template<typename DType> MSHADOW_XINLINE \
+DType name(DType a) { \
+  return DType(::name##f(static_cast<float>(a))); \
```
Comment (larroy) [inline src/operator/math_functions-inl.h]: I would add a comment regarding the intention of these and the conversion to float. Is not very clear to the reader.

Comment (mseeger) [reply src/operator/math_functions-inl.h]: Done

**Comment by piiswrong** [comment]
Could you add a test for these math ops and check them against numpy with all data types? We should try to match numpy with as high precision as possible.

Sorry for asking more work from you. But since we are looking into this now might as well get it right.

**Comment by asmushetzel** [comment]
"Well, it is always DType in = DType out, right?"

It's likely the best we can do.

**Comment by mseeger** [comment]
Hello,

I can add tests for operators which are callable by users. I'd do:
- forward: comparison Numpy, for float32, float64. It is not clear to me whether NumPy does
   computations in float32, or does not internally use float64 in any case
- finite difference tests in float64. It is our experience that such tests do not work with float32

I don't think I can test things for other DType, because NumPy only does things for float32 and float64.

@piiswrong : Do you have other DType in mind? I am not familiar with these exotic "half" types. What is important for them, is that cast from/to float is implemented.


I can add the following tests:
- Forward

**Comment by piiswrong** [comment]
```
>>> import numpy as np
>>> a = np.ones((10,), dtype=np.float16)
>>> np.exp(a)
array([ 2.71875,  2.71875,  2.71875,  2.71875,  2.71875,  2.71875,
        2.71875,  2.71875,  2.71875,  2.71875], dtype=float16)
```

numpy supports float16 exp. The behavior on int types are more complicated. Numpy does type promotion:
```
>>> a
array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int8)
>>> np.exp(a)
array([ 2.71875,  2.71875,  2.71875,  2.71875,  2.71875,  2.71875,
        2.71875,  2.71875,  2.71875,  2.71875], dtype=float16)
```

**Comment by mseeger** [comment]
@piiswrong: There is a pretty bad bug. The gradients for log10 and log2 are wrong, they are set to be 1/x in the current code, and in all tests. I'll fix this along with this PR.

**Comment by mseeger** [comment]
OK, I have now added comprehensive unit tests for unary and binary operators, and DType float16, float32, float64.

Accuracies seem OK to me. We cannot expect much for float16.

If you want better accuracy for float16, the backward expressions have to be rewritten to convert to float32 internally. But this is a bit artificial.

**Comment by piiswrong** [comment]
Its fine as long as we match numpy

**Comment by mseeger** [comment]
@piiswrong : You switched the base branch to dev.
Shall I continue to work on this to get it to master? I suppose that some unit tests may still fail.

It is quite tricky to get the accuracies right. There are unfortunately small differences on all the different target hardwares.
